lotman/lotman_ui.go

```diff
@@ -0,0 +1,573 @@
+			return http.ErrUseLastResponse
+		},
+	}
+
+	req, err := http.NewRequest("GET", directorUrl.String(), nil)
+	if err != nil {
+		return false, errors.Wrap(err, errMsgPrefix+"the director request could not be created")
+	}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return false, errors.Wrapf(err, errMsgPrefix+"the director couldn't be queried for path %s", path)
+	}
+
+	// Check the response code, make sure it's not in the error ranges (400-500)
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		return false, errors.Errorf(errMsgPrefix+"the director returned a bad status for path %s: %s", path, resp.Status)
+	}
+
+	// Get the namespace from the X-Pelican-Namespace header
+	namespaceHeader := resp.Header.Values("X-Pelican-Namespace")
+	xPelicanNamespaceMap := utils.HeaderParser(namespaceHeader[0])
+	namespace := xPelicanNamespaceMap["namespace"]
+
+	// Get the issuer URL for that namespace
+	nsIssuerUrl, err := server_utils.GetNSIssuerURL(namespace)
+	if err != nil {
+		return false, errors.Wrapf(err, errMsgPrefix+"no issuer could be found for namespace %s", namespace)
+	}
+	lot.Owner = nsIssuerUrl
+
+	return true, nil
+}
+
+func VerifyDeleteLotToken(lotName string, strToken string) (bool, error) {
+	// Get all parents of the lot, which we use to determine owners who can modify the lot itself (as opposed
+	// to the data in the lot).
+	tokenApproved := false
+
+	log.Debugf("Attempting to delete lot %s", lotName)
+	authzCallers, err := GetAuthorizedCallers(lotName)
+	if err != nil {
+		return false, errors.Wrap(err, "Error getting authorized callers")
+	}
+
+	ownerSigned, tok, err := tokenSignedByAuthorizedCaller(strToken, authzCallers)
+	if err != nil {
+		return false, errors.Wrap(err, "Error verifying token is appropriately signed")
+	}
+	if !ownerSigned {
+		return false, errors.New("Token not signed by any of the owners of any parent lot")
+	}
+
+	// We've determined the token is signed by someone we like, now to check that it has the correct lot.delete permission!
+	scope_any, present := (*tok).Get("scope")
+	if !present {
+		return false, errors.New("no scope claim in token")
+	}
+	scope, ok := scope_any.(string)
+	if !ok {
+		return false, errors.New("scope claim in token is not string-valued")
+	}
+	scopes := strings.Split(scope, " ")
+	for _, scope := range scopes {
+		if scope == token_scopes.Lot_Delete.String() {
+			tokenApproved = true
+			break
+		}
+	}
+
+	if !tokenApproved {
+		return false, errors.New("The token was correctly signed but did not possess the necessary lot.delete scope")
+	}
+
+	return true, nil
+}
+
+func VerifyUpdateLotToken(lot *LotUpdate, strToken string) (bool, error) {
+	tokenApproved := false
+	log.Debugf("Attempting to update lot %s", lot.LotName)
+
+	// Since we're updating the lot, assume it already exists (lotman will yell if it doesn't).
+	// Then we can make sure the token is signed by any owners of the lot's parents
+
+	authzCallers, err := GetAuthorizedCallers(lot.LotName)
```

Comment:
Just to confirm. There will never be a case where any owners of a lot's parents *wouldn't* be an authorize caller and any non-parent own *would*?

And I'm assuming the root lot is its own parent?

Reply:
Correct on both counts.